kernel/gic/src/mod.rs

```diff
@@ -0,0 +1,310 @@
+use bitflags::bitflags;
+
+/// Physical addresses of the CPU and Distributor
+/// interfaces as exposed by the qemu "virt" VM.
+pub mod qemu_virt_addrs {
+    use super::*;
+
+    pub const GICD: PhysicalAddress = PhysicalAddress::new_canonical(0x08000000);
+    pub const GICC: PhysicalAddress = PhysicalAddress::new_canonical(0x08010000);
+    pub const GICR: PhysicalAddress = PhysicalAddress::new_canonical(0x080A0000);
+}
+
+/// Boolean
+pub type Enabled = bool;
+
+/// 24-bit unsigned integer
+///
+/// An u32 is used because there is no u24.
+pub type IntNumber = u32;
+
+/// 8-bit unsigned integer
+pub type Priority = u8;
+
+bitflags! {
+    /// The legacy (GICv2) way of specifying
+    /// multiple target CPUs
+    pub struct TargetList: u8 {
+        const CPU_0 = 1 << 0;
+        const CPU_1 = 1 << 1;
+        const CPU_2 = 1 << 2;
+        const CPU_3 = 1 << 3;
+        const CPU_4 = 1 << 4;
+        const CPU_5 = 1 << 5;
+        const CPU_6 = 1 << 6;
+        const CPU_7 = 1 << 7;
+        const ALL_CPUS = u8::MAX;
+    }
+}
+
+pub enum TargetCpu {
+    /// That interrupt must be handled by
+    /// a specific PE in the system
+    Specific(u32),
+    /// That interrupt can be handled by
+    /// any PE that is not busy with another,
+    /// more important task
+    AnyCpuAvailable,
+    GICv2TargetList(TargetList),
+}
+
+// = 4
+const U32BYTES: usize = core::mem::size_of::<u32>();
+// = 32
+const U32BITS: usize = U32BYTES * 8;
+
+#[repr(C)]
+#[derive(zerocopy::FromBytes)]
+pub struct GicMappedPage {
+    inner: [u32; 0x400],
+}
+
+impl GicMappedPage {
+    fn read_volatile(&self, index: usize) -> u32 {
+        unsafe { (&self.inner[index] as *const u32).read_volatile() }
+    }
+
+    fn write_volatile(&mut self, index: usize, value: u32) {
+        unsafe { (&mut self.inner[index] as *mut u32).write_volatile(value) }
+    }
+
+    // Reads one slot of an array spanning across
+    // multiple u32s.
+    //
+    // - `int` is the index
+    // - `offset` tells the beginning of the array
+    // - `INTS_PER_U32` = how many array slots per u32 in this array
+    fn read_array_volatile<const INTS_PER_U32: usize>(&self, offset: usize, int: IntNumber) -> u32 {
+        let int = int as usize;
+        let bits_per_int: usize = U32BITS / INTS_PER_U32;
+        let mask: u32 = u32::MAX >> (U32BITS - bits_per_int);
+
+        let offset = offset + (int / INTS_PER_U32);
+        let reg_index = int & (INTS_PER_U32 - 1);
+        let shift = reg_index * bits_per_int;
+
+        let reg = self.read_volatile(offset);
+        (reg >> shift) & mask
+    }
+
+    // Writes one slot of an array spanning across
+    // multiple u32s.
+    //
+    // - `int` is the index
+    // - `offset` tells the beginning of the array
+    // - `INTS_PER_U32` = how many array slots per u32 in this array
+    // - `value` is the value to write
+    fn write_array_volatile<const INTS_PER_U32: usize>(&mut self, offset: usize, int: IntNumber, value: u32) {
```

Comment:
why do these functions take a const generic argument? isn't the value always `32`? In general i find these 2 functions quite confusing.

Reply:
No, sometimes it's not 32, for instance in dist_interface::get_spi_target it's 4

Reply:
ok, good to know that the const arg is needed.

I think perhaps the first line of the comment is confusing. It writes one "slot" of an array, which implies one element, but then it goes on to say that something spans multiple `u32`s. Does the element itself span multiple `u32`s, or does the whole array span multiple `u32`s`, or do you interpret the contents of the array differently based on the arguments, meaning that it's not really an array with a standard fixed layout? 

In either case you're just reading or writing a single `u32`, so the concept of spanning multiple `u32`s` doesn't really make sense, unless it's an unaligned read/write that touches different parts of two adjacent `u32`s?

Oh, and I also just noticed you're using a code comment `//` instead of a doc comment `///` with 3 slashes.

Reply:
> In either case you're just reading or writing a single `u32`, so the concept of spanning multiple `u32`s`doesn't really make sense, unless it's an unaligned read/write that touches different parts of two adjacent`u32`s?

The maximum item size is 32 bits, and the items are always aligned to 2**n bits. The array spans multiple adjacent u32s but there is always a integer number of items in a single u32.

> Oh, and I also just noticed you're using a code comment `//` instead of a doc comment `///` with 3 slashes.

I did that intentionally as it's a private function.